File: oauth-python/oauth.py
#   Copyright 2025 Google, LLC
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

"""oauth"""

import os
import json
import urllib.parse
import threading
import time
import base64
import hashlib
import secrets
import pprint
from typing import Dict, Optional, Tuple, Any, List
import webbrowser
from http.server import BaseHTTPRequestHandler, HTTPServer
import requests
from requests_oauthlib import OAuth2Session

# --- Configuration ---
CLIENT_ID: str = "oauth2python"
LOOKER_URL: str = "https://sandbox.looker-devrel.com"
AUTHORIZATION_BASE_URL: str = f"{LOOKER_URL}/auth"
LOOKER_API_URL: str = "https://sandbox.looker-devrel.com"
TOKEN_URL: str = f"{LOOKER_API_URL}/api/token"
REDIRECT_PORT: int = 8080  # Define port before using it in REDIRECT_URI
REDIRECT_URI: str = (
  f"http://localhost:{REDIRECT_PORT}/callback"  # Must be registered with your provider
)
SCOPES: List[str] = ["cors_api"]  # Your desired scopes
TOKEN_FILE: str = "oauth_tokens.json"


# --- PKCE Generation ---
def generate_pkce_pair() -> Tuple[str, str]:
  """Generates a PKCE code_verifier and code_challenge."""
  # code_verifier: a high-entropy cryptographically random string from 43 to 128 octets long
  # (before base64url-encoding).
  # RFC 7636 suggests 32 octets for a good random source.
  code_verifier: str = secrets.token_urlsafe(
    96
  )  # Generates 128 characters safe for URL

  # code_challenge_method: S256 (SHA256 hash)
  # code_challenge: BASE64URL-encode(SHA256(ASCII(code_verifier)))
  s256: bytes = hashlib.sha256(code_verifier.encode("utf-8")).digest()
  code_challenge: str = base64.urlsafe_b64encode(s256).decode("utf-8").rstrip("=")

  return code_verifier, code_challenge


# --- Local HTTP Server for Redirect ---
class OAuthCallbackHandler(BaseHTTPRequestHandler):
  """OAuthCallbackHandler class"""

  server: "OAuthCallbackServer"  # Forward reference for type hint

  # pylint: disable-next=invalid-name
  def do_GET(self) -> None:
    """do_GET called on GET"""
    parsed_url: urllib.parse.ParseResult = urllib.parse.urlparse(self.path)
    query_params: Dict[str, List[str]] = urllib.parse.parse_qs(parsed_url.query)

    if "code" in query_params:
      auth_code: str = query_params["code"][0]
      self.server.auth_code = auth_code
      self.send_response(200)
      self.send_header("Content-type", "text/html")
      self.end_headers()
      self.wfile.write(b"Authorization successful! You can close this tab.")
    elif "error" in query_params:
      error: str = query_params["error"][0]
      error_description: str = query_params.get("error_description", [""])[0]
      self.server.auth_error = f"Error: {error}, Description: {error_description}"
      self.send_response(400)
      self.send_header("Content-type", "text/html")
      self.end_headers()
      self.wfile.write(
        f"Authorization failed: {error} - {error_description}".encode("utf-8")
      )
    else:
      self.send_response(400)
      self.send_header("Content-type", "text/html")
      self.end_headers()
      self.wfile.write(b"Authorization failed or code not found.")
    # Trigger server shutdown (non-blocking)
    threading.Thread(target=self.server.shutdown).start()


class OAuthCallbackServer(HTTPServer):
  """OAuthCallbackServer"""

  auth_code: Optional[str]
  auth_error: Optional[str]

  def __init__(self, server_address: Tuple[str, int], RequestHandlerClass: Type[BaseHTTPRequestHandler]):
    super().__init__(server_address, RequestHandlerClass)
    self.auth_code = None
    self.auth_error = None


def start_local_server_and_wait_for_code() -> OAuthCallbackServer:
  """start_local_server_and_wait_for_code"""

  host: Optional[str] = urllib.parse.urlparse(REDIRECT_URI).hostname
  # Ensure host is not None, though practically it should always be there for http/https URLs
  if host is None:
    raise ValueError("Could not parse hostname from REDIRECT_URI")
  port: int = REDIRECT_PORT
  server_address: Tuple[str, int] = (host, port)
  httpd = OAuthCallbackServer(server_address, OAuthCallbackHandler)
  print(f"Starting local server on {REDIRECT_URI} to catch redirect...")
  server_thread: threading.Thread = threading.Thread(target=httpd.serve_forever)
  server_thread.start()
  # Give a small delay for the server to start
  time.sleep(1)
  return httpd


# --- Token Management ---
def load_tokens() -> Optional[Dict[str, Any]]:
  """load_tokens"""
  if os.path.exists(TOKEN_FILE):
    with open(TOKEN_FILE, "r", encoding='utf-8') as f:
      tokens: Dict[str, Any] = json.load(f)
      return tokens
  return None


def save_tokens(tokens: Dict[str, Any]) -> None:
  """save_tokens"""
  # Add expiry time for future checks if not present
  if "expires_in" in tokens and "expires_at" not in tokens:
    tokens["expires_at"] = time.time() + tokens["expires_in"]
  with open(TOKEN_FILE, "w", encoding='utf-8') as f:
    json.dump(tokens, f, indent=4)
  print(f"Tokens saved to {TOKEN_FILE}")

def refresh_access_token(
  # pylint: disable-next=unused-argument
  oauth: OAuth2Session, tokens: Dict[str, Any]
) -> Optional[Dict[str, Any]]:
  """refresh_access_token"""

  print("Access token expired, refreshing...")
  try:
    data: Dict[str, str] = {
      "grant_type": "refresh_token",
      "refresh_token": str(
        tokens["refresh_token"]
      ),  # Ensure refresh_token is string
      "client_id": CLIENT_ID,
      # 'client_secret': CLIENT_SECRET # For public clients, this is usually NOT included
    }
    response: requests.Response = requests.post(TOKEN_URL, data=data, timeout=300)
    response.raise_for_status()  # Raise an exception for HTTP errors
    new_tokens: Dict[str, Any] = response.json()
    save_tokens(new_tokens)
    print("Token refreshed successfully.")
    return new_tokens
  except requests.exceptions.RequestException as e:
    print(f"Error refreshing token: {e}")
    # Consider revoking the refresh token if it's consistently failing
    return None


def get_authorized_session() -> Optional[OAuth2Session]:
  """get_authorized_session"""

  tokens: Optional[Dict[str, Any]] = load_tokens()
  oauth: OAuth2Session = OAuth2Session(
    CLIENT_ID, scope=SCOPES, redirect_uri=REDIRECT_URI
  )

  if tokens:
    oauth.token = tokens
    # Check if access token is expired
    # Ensure 'expires_at' is present and is a number for comparison
    expires_at = tokens.get("expires_at")
    if (
      isinstance(expires_at, (int, float)) and expires_at < time.time() - 300
      ):  # Give a 5 min buffer
      print("Access token is expired.")
      refreshed_tokens = refresh_access_token(oauth, tokens)
      if not refreshed_tokens:
        print("Failed to refresh token. Will attempt full re-authorization.")
        tokens = None  # Force full re-authorization
      else:
        tokens = refreshed_tokens
        oauth.token = tokens  # Update session with new tokens
    elif "access_token" in tokens:
      print("Using existing valid access token.")
      return oauth

  if not tokens:  # No tokens, or refresh failed, initiate full authorization
    print("Initiating new authorization flow...")
    code_verifier, code_challenge = generate_pkce_pair()
    # print(f"Generated code_verifier (keep secret): {code_verifier}")
    # For debugging, DO NOT LOG IN PRODUCTION

    # Create the authorization URL with PKCE parameters
    authorization_url: str
    # pylint: disable-next=unused-variable
    state: str
    authorization_url, state = oauth.authorization_url(
      AUTHORIZATION_BASE_URL,
      code_challenge=code_challenge,
      code_challenge_method="S256",
    )

    print("\nPlease go to this URL to authorize your application:")
    print(authorization_url)
    webbrowser.open(authorization_url)

    # Start local server to capture the code
    httpd: OAuthCallbackServer = start_local_server_and_wait_for_code()

    # Wait for the server to finish serving the request (i.e., get the code)
    # We need to wait for the server thread to shut down before continuing
    # This is handled by threading.Thread(target=self.server.shutdown).start() in the handler
    httpd.serve_forever()  # This will block until shutdown() is called
    httpd.server_close()  # Ensure the socket is closed

    auth_code: Optional[str] = httpd.auth_code
    auth_error: Optional[str] = httpd.auth_error

    if auth_error:
      print(f"Authorization failed: {auth_error}")
      return None

    if not auth_code:
      print(
          "Failed to receive authorization code."
      )  # Removed "within timeout" as serve_forever blocks.
      return None

    print("\nReceived authorization code, exchanging for tokens...")
    try:
      token_response: Dict[str, Any] = oauth.fetch_token(
        token_url=TOKEN_URL,
        code=auth_code,
        code_verifier=code_verifier,  # Crucial for PKCE
        client_id=CLIENT_ID,
        # Client ID is often included in token exchange for public clients
        # client_secret=CLIENT_SECRET # For public clients, this is usually NOT included
        include_client_id=True,  # Crucial for PKCE
      )
      save_tokens(token_response)
      print("Tokens obtained and saved successfully.")
      return oauth
    except requests.exceptions.RequestException as e:
      print(f"Error exchanging code for tokens: {e}")
      if hasattr(e, "response") and e.response is not None:
        print(f"Response content: {e.response.text}")
      return None
  return oauth  # Should return the session if tokens were valid from the start and not expired


# --- Main Program Logic ---
def main() -> None:
  """Orchestrates the OAuth 2.0 authorization flow and makes a sample API call."""
  session: Optional[OAuth2Session] = get_authorized_session()
  if session:
    # Example API call (replace with your actual API endpoint)
    # The token is available as session.token
    try:
      api_url: str = f"{LOOKER_API_URL}/api/4.0/user?fields=id,display_name,email"
      response: requests.Response = session.get(api_url)
      response.raise_for_status()  # Raise an exception for HTTP errors
      print("API call successful!")
      api_data: Any = response.json()
      pprint.pprint(api_data)
    except requests.exceptions.RequestException as e:
      print(f"API call failed: {e}")
  else:
    print("Could not authorize application.")


if __name__ == "__main__":
  main()

File: oauth-ruby/oauth.rb
#   Copyright 2025 Google, LLC
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

require 'webrick'
require 'net/http'
require 'json'
require 'uri'
require 'pkce_challenge'
require 'launchy'

# --- Configuration ---
CLIENT_ID = 'oauth2ruby'
LOOKER_URL = 'https://sandbox.looker-devrel.com'
AUTHORIZATION_BASE_URL = "#{LOOKER_URL}/auth"
LOOKER_API_URL = 'https://sandbox.looker-devrel.com'
TOKEN_URL = "#{LOOKER_API_URL}/api/token"
REDIRECT_PORT = 8080
REDIRECT_URI = "http://localhost:#{REDIRECT_PORT}/callback"
SCOPE = 'cors_api'
TOKEN_FILE = 'oauth_tokens.json'
# --- End Configuration ---


# Use a simple in-memory hash to store the code_verifier between requests.
# In a real production app, you would use a more robust session store.
$session_store = {}

# Helper method for URL-safe Base64 encoding
def base64_url_encode(str)
  Base64.urlsafe_encode64(str, padding: false)
end

# Generate the PKCE code verifier and challenge
def generate_pkce_pair()
  pkce = PkceChallenge.challenge(char_length: 128)
  code_verifier = pkce.code_verifier
  code_challenge = pkce.code_challenge

  return code_verifier, code_challenge
end

def load_tokens()
  return nil unless File.exist?(TOKEN_FILE)
  s = File.stat(TOKEN_FILE)
  if !(s.mode.to_s(8)[3..5] == "600")
    say_error "#{TOKEN_FILE} mode is #{s.mode.to_s(8)[3..5]}. It must be 600. Ignoring."
    return nil
  end
  token_data = nil
  file = nil
  begin
    file = File.open(TOKEN_FILE)
    token_data = JSON.parse(file.read,{:symbolize_names => true})
  ensure
    file.close if file
  end
  token_data
end

def save_tokens(tokens)
  if tokens[:expires_in] and not tokens[:expires_at]
    tokens[:expires_at] = Time.now + tokens[:expires_in]
  end
  file = nil
  begin
    file = File.new(TOKEN_FILE, "wt")
    file.chmod(0600)
    file.write JSON.pretty_generate(tokens)
  ensure
    file.close if file
  end
end

def refresh_access_token(tokens)
  puts "Access token expired, refreshing..."
  begin
    data = {
      :grant_type => 'refresh_token',
      :refresh_token => tokens[:refresh_token],
      :client_id => CLIENT_ID
    }

    token_uri = URI(TOKEN_URL)

    http = Net::HTTP.new(token_uri.host, token_uri.port)
    http.use_ssl = true
    request = Net::HTTP::Post.new(token_uri.request_uri)
    request.set_form_data(data)

    token_response = http.request(request)
    new_tokens = JSON.parse(token_response.body,{:symbolize_names => true})
    if new_tokens[:error]
      puts "Error refreshing tokens"
      puts new_tokens
      return nil
    end
    save_tokens(new_tokens)
    puts "Token refreshed successfully"
    return new_tokens
  rescue Exception => e
    puts "Error refreshing token: #{e}"
    return nil
  end
end

def get_authorized_session()
  tokens = load_tokens()

  if tokens and tokens[:error]
    tokens = nil
  end

  if tokens
    if tokens[:expires_at]
      (day, time, tz) = tokens[:expires_at].split(' ')
      day_parts = day.split('-')
      time_parts = time.split(':')
      date_time_parts = day_parts + time_parts + [tz]
      expiration = Time.new(*date_time_parts)
      if expiration < (Time.now + 300)
        puts "Access token is expired"
        tokens = refresh_access_token(tokens)
      end
      if not tokens
        puts "Failed to refresh token. Will attempt full reauthorization"
        tokens = nil
      else
        # Update session with new tokens
        $session_store[:access_token] = tokens[:access_token]
      end
    elsif tokens[:access_token]
      puts "Using existing access token."
      $session_store[:access_token] = tokens[:access_token]
    end
  end

  if !tokens
    puts "Initializing new Authorization flow."
    # Create the authentication request
    code_verifier, code_challenge = generate_pkce_pair()
    $session_store[:code_verifier] = code_verifier
    $session_store[:state] = SecureRandom.urlsafe_base64(64)

    # Build the authorization URL
    auth_params = {
      response_type: 'code',
      client_id: CLIENT_ID,
      redirect_uri: REDIRECT_URI,
      scope: SCOPE,
      state: $session_store[:state], # A random string for security
      code_challenge_method: 'S256',
      code_challenge: code_challenge
    }

    auth_uri = URI(AUTHORIZATION_BASE_URL)
    auth_uri.query = URI.encode_www_form(auth_params)

    puts "🚀 Starting server on http://localhost:#{REDIRECT_PORT}"
    server = configure_server(auth_uri)
    server.start

    if $session_store[:auth_error]
      puts "Authorization failed: #{$session_store[:auth_error]}"
      return nil
    end

    if not $session_store[:auth_code]
      puts "Failed to receive authorization code within timeout."
      return nil
    end

    puts "Received Authoriztion code, exchanging for tokens..."
    begin
      token_uri = URI(TOKEN_URL)

      token_request_body = {
        grant_type: 'authorization_code',
        client_id: CLIENT_ID,
        redirect_uri: REDIRECT_URI,
        code: $session_store[:auth_code],
        code_verifier: $session_store[:code_verifier] # Send the original verifier
      }

      # Make the POST request to the token endpoint
      http = Net::HTTP.new(token_uri.host, token_uri.port)
      http.use_ssl = true
      request = Net::HTTP::Post.new(token_uri.request_uri)
      request.set_form_data(token_request_body)

      token_response = http.request(request)
      token_data = JSON.parse(token_response.body,{:symbolize_names => true})

      save_tokens(token_data)

      if token_data[:error]
        puts "Error getting token:#{JSON.pretty_generate(token_data)}"
      end

      $session_store[:access_token] = token_data[:access_token]
    end
  end
end

def configure_server(auth_uri)
  # Create a new WEBrick server
  server = WEBrick::HTTPServer.new(
    Port: REDIRECT_PORT,
    StartCallback: -> {
      # direct the user's browser to the Looker login page
      Launchy.open(auth_uri.to_s)
    }
  )


  # Mount a handler for the callback URI ('/callback')
  # This is where Looker redirects after the user authenticates.
  server.mount_proc '/callback' do |req, res|
    # Extract the authorization code from the query parameters
    code = req.query['code']
    error = req.query['error']
    state = req.query['state']

    if code and state == $session_store[:state]
      $session_store[:auth_code] = code
      res.status = 200
      res.body = "Authorization successful! You can close this tab."
    elsif error
      error_description = req.query.get('error_description', '')
      $session_store[:auth_error] = "Error: #{error}, Description: #{error_description}"
      res.status = 400
      res.body = "Error: Authorization failed: #{error} - #{error_description}"
    else
      res.status = 400
      res.body = "Error: Authorization failed or code not found."
    end
    server.shutdown
  end

  return server
end

get_authorized_session()

access_token = $session_store[:access_token]

# Use the access token to make an authenticated API call
user_api_uri = URI("#{LOOKER_API_URL}/api/4.0/user?fields=id,display_name,email")

user_request = Net::HTTP::Get.new(user_api_uri.request_uri)
user_request['Authorization'] = "Bearer #{access_token}"

user_http = Net::HTTP.new(user_api_uri.host, user_api_uri.port)
user_http.use_ssl = true
user_response = user_http.request(user_request)
user_data = JSON.parse(user_response.body)

# Display the results
puts """
✅ Authentication Successful!
Your access token has been retrieved and used to call the API.

Access Token:
#{access_token[0..30]}...

API Response from /api/4.0/user:
#{JSON.pretty_generate(user_data)}
"""
